### ISSUE TITLE
Cannot connect to servers with bad TLS configs - provide more options to connectWithTLS function

### ISSUE THREAD

**Issue by AdamHarries** [barrucadu/irc-client#22]
I'm looking to use this library to connect to a friends server which unfortunately has a slightly broken TLS setup, but mandates TLS for all clients. On my desktop client (weechat), I can just set `ssl_verify = off` in my config, but there's no way to set options such as this with irc-client.

Would it be possible to expose more connection options to the user? 

At present, trying to connect to a server with a bad config gives:

`HandshakeFailed (Error_Protocol ("certificate rejected: [NameMismatch \"<redacted>\"]",True,CertificateUnknown))`

**Comment by barrucadu** [comment]
The certificate validation comes from `ircTLSClient` in [irc-conduit](https://github.com/barrucadu/irc-conduit):

~~~haskell
-- |Like 'ircClient', but with TLS.
ircTLSClient :: Int -> ByteString -> IO () -> Consumer (Either ByteString IrcEvent) IO () -> Producer IO IrcMessage -> IO ()
ircTLSClient port host = ircWithConn $ runTLSClient $ mangle $ tlsClientConfig port host
  where
    -- Override the certificate validation and allowed ciphers.
    mangle tlsSettings = tlsSettings
                         { tlsClientTLSSettings = TLSSettings cpara
                           { clientHooks = (clientHooks cpara)
                             { onServerCertificate = validate }
                           , clientSupported = (clientSupported cpara)
                             { supportedVersions = [TLS12, TLS11, TLS10]
                             , supportedCiphers = ciphersuite_all }}}
      where
        cpara = defaultParamsClient (unpack $ decodeUtf8 host) ""

        -- Make the TLS certificate validation a bit more generous. In
        -- particular, allow self-signed certificates.
        validate cs vc sid cc = do
          -- First validate with the standard function
          res <- (onServerCertificate $ clientHooks cpara) cs vc sid cc
          -- Then strip out non-issues
          return $ filter (`notElem` [UnknownCA, SelfSigned]) res
~~~

Unfortunately I don't expose enough internals to actually define your own alternative function and then use that instead... I'll add something to the API to allow you to pass in your own SSL config, and also expose the currently-hidden `Network.IRC.Client.Internal` module.

**Comment by barrucadu** [comment]
I have just released version 0.4.4.0, which has two functions which should interest you:

- `connectWithTLSConfig :: MonadIO m => TLSClientConfig -> NominalDiffTime -> m (ConnectionConfig s)`

    Which allows you to specify the *entire* TLS config, you can even go nuts and specify the ciphersuite if you want.

- `connectWithTLSVerify :: MonadIO m => (CertificateStore -> ValidationCache -> ServiceID -> CertificateChain -> IO [FailedReason]) -> ByteString -> Int -> NominalDiffTime -> m (ConnectionConfig s)`

    Which allows you to give the certificate verification function. To turn off verification entirely, you can use the function `\_ _ _ _ -> pure []`.

**Comment by AdamHarries** [comment]
Thanks - I think that perfectly fixes my problem!
